File: frontend/src/music/chordRender.ts
/**
 * Chord rendering utilities for converting chord events to playable notes.
 *
 * NOTE: This frontend renderer should match the backend chord_patterns.py renderer
 * for consistency. For production, consider using the backend renderer via API
 * or sharing the algorithm via tests.
 */

import type { ChordEvent } from "../types";

const PPQ = 480; // Pulses per quarter note

const DEBUG =
  typeof window !== "undefined" &&
  (localStorage.getItem("midinecromancer:debug:chordRender") === "true" || import.meta.env.DEV);

function debugLog(...args: any[]) {
  if (DEBUG) {
    console.log("[chordRender]", ...args);
  }
}

/**
 * Convert a chord symbol to MIDI note numbers.
 * Improved parser that handles sharps, flats, and various chord qualities.
 * @param chordName - Chord name (e.g., "C", "Am", "G7", "F#m", "Bbmaj7")
 * @param romanNumeral - Optional roman numeral (e.g., "I", "vi", "V7") - more reliable
 * @param tonic - Key tonic (e.g., "C", "G#")
 * @param mode - Mode (e.g., "ionian", "aeolian")
 * @param octave - Base octave (default 4)
 * @returns Array of MIDI note numbers
 */
function chordNameToNotes(
  chordName: string,
  tonic: string,
  mode: string,
  octave = 4,
  romanNumeral?: string
): number[] {
  // Map note names to semitones from C
  const noteMap: Record<string, number> = {
    C: 0,
    "C#": 1,
    Db: 1,
    D: 2,
    "D#": 3,
    Eb: 3,
    E: 4,
    F: 5,
    "F#": 6,
    Gb: 6,
    G: 7,
    "G#": 8,
    Ab: 8,
    A: 9,
    "A#": 10,
    Bb: 10,
    B: 11,
  };

  // If we have a roman numeral, use it for more reliable pitch calculation
  if (romanNumeral) {
    return chordNameToNotesFromRoman(romanNumeral, tonic, mode, octave);
  }

  // Parse chord name - handle sharps/flats in root note
  // Examples: "C", "C#", "F#m", "Bbm", "G#maj7", "Ab7"
  const chordNameUpper = chordName.trim();

  // Extract root note (may include # or b)
  let rootNote = "";
  let i = 0;

  // Handle single letter or letter + sharp/flat
  if (chordNameUpper[i] && /[A-G]/.test(chordNameUpper[i])) {
    rootNote += chordNameUpper[i];
    i++;
    // Check for # or b
    if (chordNameUpper[i] === "#" || chordNameUpper[i] === "b") {
      rootNote += chordNameUpper[i];
      i++;
    }
  }

  if (!rootNote) {
    console.warn(`Could not parse root note from chord: ${chordName}`);
    return [];
  }

  // Parse quality
  const remainder = chordNameUpper.slice(i);
  const isMinor = remainder.includes("m") || remainder.includes("min");
  const is7th = remainder.includes("7");
  const isMajor7 = remainder.includes("maj7") || remainder.includes("M7");
  const isDim = remainder.includes("dim");
  const isAug = remainder.includes("aug");

  const rootSemitone = noteMap[rootNote];
  if (rootSemitone === undefined) {
    console.warn(`Unknown root note: ${rootNote} in chord ${chordName}`);
    return [];
  }

  const rootMidi = (octave + 1) * 12 + rootSemitone;

  // Build chord intervals
  const intervals: number[] = [0]; // Root

  if (isDim) {
    intervals.push(3); // Minor third
    intervals.push(6); // Diminished fifth
  } else if (isAug) {
    intervals.push(4); // Major third
    intervals.push(8); // Augmented fifth
  } else if (isMinor) {
    intervals.push(3); // Minor third
    intervals.push(7); // Perfect fifth
  } else {
    intervals.push(4); // Major third
    intervals.push(7); // Perfect fifth
  }

  if (is7th) {
    if (isMajor7) {
      intervals.push(11); // Major 7th
    } else {
      intervals.push(10); // Minor 7th (dominant 7th)
    }
  }

  return intervals.map((interval) => rootMidi + interval);
}

/**
 * Convert roman numeral to MIDI notes using scale degrees.
 * This is more reliable than parsing chord names.
 */
function chordNameToNotesFromRoman(
  roman: string,
  tonic: string,
  mode: string,
  octave = 4
): number[] {
  // Map roman numerals to scale degrees
  const romanToDegree: Record<string, number> = {
    I: 1,
    II: 2,
    III: 3,
    IV: 4,
    V: 5,
    VI: 6,
    VII: 7,
    i: 1,
    ii: 2,
    iii: 3,
    iv: 4,
    v: 5,
    vi: 6,
    vii: 7,
  };

  // Extract base roman (remove 7, sus, etc.)
  const baseRoman = roman.replace(/7|sus|dim|aug/g, "").trim();
  const degree = romanToDegree[baseRoman] || 1;
  const is7th = roman.includes("7");
  const isMinor = baseRoman === baseRoman.toLowerCase() || baseRoman.includes("m");

  // Map tonic to semitone
  const noteMap: Record<string, number> = {
    C: 0,
    "C#": 1,
    Db: 1,
    D: 2,
    "D#": 3,
    Eb: 3,
    E: 4,
    F: 5,
    "F#": 6,
    Gb: 6,
    G: 7,
    "G#": 8,
    Ab: 8,
    A: 9,
    "A#": 10,
    Bb: 10,
    B: 11,
  };

  const tonicSemitone = noteMap[tonic] || 0;

  // Mode intervals (semitones from tonic)
  const modeIntervals: Record<string, number[]> = {
    ionian: [0, 2, 4, 5, 7, 9, 11],
    aeolian: [0, 2, 3, 5, 7, 8, 10],
    dorian: [0, 2, 3, 5, 7, 9, 10],
    phrygian: [0, 1, 3, 5, 7, 8, 10],
    lydian: [0, 2, 4, 6, 7, 9, 11],
    mixolydian: [0, 2, 4, 5, 7, 9, 10],
    locrian: [0, 1, 3, 5, 6, 8, 10],
  };

  const intervals = modeIntervals[mode] || modeIntervals.ionian;
  const degreeIdx = (degree - 1) % 7;
  const rootSemitone = (tonicSemitone + intervals[degreeIdx]) % 12;
  const rootMidi = (octave + 1) * 12 + rootSemitone;

  // Build chord notes directly from scale degrees
  const notes: number[] = [];

  // Root
  const rootNote = (octave + 1) * 12 + rootSemitone;
  notes.push(rootNote);

  // Third (degree + 2)
  const thirdIdx = (degreeIdx + 2) % 7;
  const thirdSemitone = (tonicSemitone + intervals[thirdIdx]) % 12;
  notes.push((octave + 1) * 12 + thirdSemitone);

  // Fifth (degree + 4)
  const fifthIdx = (degreeIdx + 4) % 7;
  const fifthSemitone = (tonicSemitone + intervals[fifthIdx]) % 12;
  notes.push((octave + 1) * 12 + fifthSemitone);

  // Seventh (degree + 6) if needed
  if (is7th) {
    const seventhIdx = (degreeIdx + 6) % 7;
    const seventhSemitone = (tonicSemitone + intervals[seventhIdx]) % 12;
    notes.push((octave + 1) * 12 + seventhSemitone);
  }

  return notes.sort((a, b) => a - b);
}

/**
 * Apply voicing to chord notes.
 * @param notes - Base chord notes
 * @param voicing - Voicing style ("root", "open", "drop2", "smooth")
 * @param inversion - Inversion number (0-3)
 * @param lowMidi - Lowest MIDI note (default 48 = C3)
 * @param highMidi - Highest MIDI note (default 72 = C5)
 * @returns Voiced chord notes
 */
function applyVoicing(
  notes: number[],
  voicing: string,
  inversion: number,
  lowMidi = 48,
  highMidi = 72
): number[] {
  if (notes.length === 0) return [];

  // Apply inversion
  let voiced = [...notes];
  for (let i = 0; i < inversion && i < notes.length; i++) {
    const root = voiced[0];
    voiced = voiced.slice(1);
    voiced.push(root + 12); // Move root up an octave
  }

  // Apply voicing style
  if (voicing === "open") {
    // Spread notes across wider range
    const spread = voiced.map((note, idx) => {
      const octaveOffset = Math.floor(idx / 3) * 12;
      return note + octaveOffset;
    });
    voiced = spread;
  } else if (voicing === "drop2") {
    // Drop second note an octave
    if (voiced.length >= 2) {
      voiced[1] -= 12;
    }
  } else if (voicing === "smooth") {
    // Keep notes close together (within one octave)
    // Already handled by inversion
  }
  // "root" voicing: keep as-is

  // Clamp to range
  return voiced
    .map((note) => {
      // Move to target range if needed
      while (note < lowMidi) note += 12;
      while (note > highMidi) note -= 12;
      return note;
    })
    .filter((note) => note >= lowMidi && note <= highMidi)
    .sort((a, b) => a - b);
}

/**
 * Render a chord event to note events for playback.
 * @param chordEvent - Chord event to render
 * @param clipStartBar - Starting bar of the clip
 * @param clipOffsetTicks - Clip offset in ticks
 * @param trackOffsetTicks - Track offset in ticks
 * @param ticksPerBar - Ticks per bar
 * @param tonic - Key tonic
 * @param mode - Mode
 * @param seed - Random seed for determinism
 * @param bpm - BPM for beat-to-tick conversion (optional, defaults to 120)
 * @returns Array of note events ready for Tone.js
 */
export function renderChordEventToNotes(
  chordEvent: ChordEvent,
  clipStartBar: number,
  clipOffsetTicks: number,
  trackOffsetTicks: number,
  ticksPerBar: number,
  tonic: string,
  mode: string,
  seed: number,
  bpm = 120
): Array<{
  pitch: number;
  start_tick: number;
  duration_tick: number;
  velocity: number;
}> {
  if (!chordEvent.is_enabled) {
    return [];
  }

  // Get base chord notes - prefer roman numeral if available (more reliable)
  const baseNotes = chordNameToNotes(
    chordEvent.chord_name,
    tonic,
    mode,
    4,
    chordEvent.roman_numeral
  );

  if (baseNotes.length === 0) {
    if (DEBUG) {
      console.warn(`[chordRender] No notes generated for chord:`, {
        chord_name: chordEvent.chord_name,
        roman: chordEvent.roman_numeral,
        tonic,
        mode,
      });
    }
    return [];
  }

  // Apply voicing
  const voicedNotes = applyVoicing(
    baseNotes,
    chordEvent.voicing || "root",
    chordEvent.inversion || 0,
    48, // C3
    72 // C5
  );

  if (voicedNotes.length === 0) {
    debugLog("No voiced notes after voicing/inversion", {
      baseNotes,
      voicing: chordEvent.voicing,
      inversion: chordEvent.inversion,
    });
    return [];
  }

  debugLog("Rendering chord", {
    chord_name: chordEvent.chord_name,
    roman: chordEvent.roman_numeral,
    baseNotes,
    voicedNotes,
    strum_beats: chordEvent.strum_beats,
    humanize_beats: chordEvent.humanize_beats,
  });

  // Calculate timing
  const baseStartTick =
    clipStartBar * ticksPerBar + chordEvent.start_tick + clipOffsetTicks + trackOffsetTicks;
  const durationTicks = chordEvent.duration_tick;
  const intensity = chordEvent.intensity || 0.85;

  // Convert strum/humanize from beats to ticks
  // Use beats fields if available, otherwise fall back to ms (for backward compatibility)
  let strumBeats = chordEvent.strum_beats ?? 0.0;
  let humanizeBeats = chordEvent.humanize_beats ?? 0.0;

  // Fallback: convert ms to beats if beats fields are not available
  if (strumBeats === 0 && chordEvent.strum_ms && chordEvent.strum_ms > 0) {
    strumBeats = (chordEvent.strum_ms / 1000.0) * (bpm / 60.0);
  }
  if (humanizeBeats === 0 && chordEvent.humanize_ms && chordEvent.humanize_ms > 0) {
    humanizeBeats = (chordEvent.humanize_ms / 1000.0) * (bpm / 60.0);
  }

  // Convert beats to ticks (1 beat = 1 quarter note = PPQ ticks)
  const ticksPerBeat = PPQ;
  const strumTicks = Math.round(strumBeats * ticksPerBeat);
  const humanizeTicks = Math.round(humanizeBeats * ticksPerBeat);

  // Deterministic random for strum order and humanization
  // Simple seeded random function
  function seededRandom(s: number): number {
    s = Math.sin(s) * 10000;
    return s - Math.floor(s);
  }

  const strumSeed = seed + chordEvent.id.charCodeAt(0);
  const humanizeSeed = seed + chordEvent.id.charCodeAt(chordEvent.id.length - 1);

  // Determine strum order
  const strumOrder = [...Array(voicedNotes.length).keys()];
  if (strumTicks > 0) {
    // Shuffle deterministically
    for (let i = strumOrder.length - 1; i > 0; i--) {
      const j = Math.floor(seededRandom(strumSeed + i) * (i + 1));
      [strumOrder[i], strumOrder[j]] = [strumOrder[j], strumOrder[i]];
    }
  }

  // Generate note events
  const noteEvents: Array<{
    pitch: number;
    start_tick: number;
    duration_tick: number;
    velocity: number;
  }> = [];

  for (let i = 0; i < strumOrder.length; i++) {
    const noteIdx = strumOrder[i];
    const pitch = voicedNotes[noteIdx];

    // Calculate start time with strum
    // Distribute notes evenly across strum duration
    let noteStart: number;
    if (strumBeats > 0 && strumOrder.length > 1) {
      const strumPosition = i / (strumOrder.length - 1);
      noteStart = baseStartTick + Math.round(strumPosition * strumTicks);
    } else {
      noteStart = baseStartTick;
    }

    // Apply humanization (clamp to chord bounds)
    if (humanizeBeats > 0) {
      const humanizeOffset = Math.round((seededRandom(humanizeSeed + i) - 0.5) * 2 * humanizeTicks);
      noteStart += humanizeOffset;
      // Clamp to chord bounds
      noteStart = Math.max(baseStartTick, Math.min(baseStartTick + durationTicks, noteStart));
    }

    // Calculate velocity with intensity
    let velocity = Math.round(100 * intensity);
    if (chordEvent.velocity_jitter > 0) {
      const jitter = Math.round(
        (seededRandom(humanizeSeed + i + 100) - 0.5) * 2 * chordEvent.velocity_jitter
      );
      velocity += jitter;
    }
    velocity = Math.max(1, Math.min(127, velocity));

    noteEvents.push({
      pitch,
      start_tick: noteStart,
      duration_tick: durationTicks,
      velocity,
    });
  }

  return noteEvents;
}

File: backend/src/midinecromancer/services/chord_render.py
"""Service for rendering chord events to notes with expressive parameters."""

import hashlib
import random
from typing import TYPE_CHECKING

from midinecromancer.music.theory import PPQ, get_chord_notes, roman_to_degree

if TYPE_CHECKING:
    from midinecromancer.models.chord_event import ChordEvent


def deterministic_seed(base_seed: int, chord_id: str, param: str) -> int:
    """Generate deterministic seed for a chord parameter."""
    hash_input = f"{base_seed}:{chord_id}:{param}".encode()
    hash_bytes = hashlib.blake2b(hash_input, digest_size=8).digest()
    return int.from_bytes(hash_bytes, byteorder="big")


def render_chord_event_to_notes(
    chord_event: "ChordEvent",
    project_context: dict,
    seed: int,
) -> list[dict]:
    """Render a single chord event to note events with expressive parameters.

    Args:
        chord_event: ChordEvent model with expressive fields
        project_context: Dict with tonic, mode, bpm, time_signature_num, time_signature_den
        seed: Base seed for determinism

    Returns:
        List of note dicts: {pitch, start_tick, duration_tick, velocity}
    """
    tonic = project_context["tonic"]
    mode = project_context["mode"]
    bpm = project_context["bpm"]

    # Get chord tones
    roman = chord_event.roman_numeral
    degree = roman_to_degree(roman)
    quality = "7th" if "7" in roman else "triad"
    chord_tones = get_chord_notes(tonic, mode, degree, quality, octave=4)

    # Apply voicing and inversion
    voicing_pitches = apply_voicing(
        chord_tones,
        chord_event.voicing,
        chord_event.inversion,
        voicing_low=48,
        voicing_high=72,
    )

    # Calculate timing
    base_start_tick = chord_event.start_tick
    duration_ticks = int(chord_event.duration_tick)
    intensity = float(chord_event.intensity)

    # Convert strum/humanize from beats to ticks
    # Use beats fields if available, otherwise fall back to ms (for backward compatibility)
    quarter_notes_per_bar = (project_context["time_signature_num"] * 4) / project_context["time_signature_den"]
    ticks_per_beat = PPQ  # 1 beat = 1 quarter note = PPQ ticks
    
    if hasattr(chord_event, "strum_beats") and chord_event.strum_beats is not None:
        strum_beats = float(chord_event.strum_beats)
    else:
        # Fallback: convert ms to beats (approximate, requires BPM)
        strum_beats = (chord_event.strum_ms / 1000.0) * (bpm / 60.0) if chord_event.strum_ms > 0 else 0.0
    
    if hasattr(chord_event, "humanize_beats") and chord_event.humanize_beats is not None:
        humanize_beats = float(chord_event.humanize_beats)
    else:
        # Fallback: convert ms to beats
        humanize_beats = (chord_event.humanize_ms / 1000.0) * (bpm / 60.0) if chord_event.humanize_ms > 0 else 0.0
    
    strum_ticks = int(strum_beats * ticks_per_beat)
    humanize_ticks = int(humanize_beats * ticks_per_beat)

    # Generate deterministic seeds for each parameter
    strum_seed = deterministic_seed(seed, str(chord_event.id), "strum")
    humanize_seed = deterministic_seed(seed, str(chord_event.id), "humanize")
    velocity_seed = deterministic_seed(seed, str(chord_event.id), "velocity")

    notes = []
    rng_strum = random.Random(strum_seed)
    rng_humanize = random.Random(humanize_seed)
    rng_velocity = random.Random(velocity_seed)

    # Determine strum order (deterministic)
    strum_order = list(range(len(voicing_pitches)))
    if strum_beats > 0:
        # Random but deterministic order
        rng_strum.shuffle(strum_order)

    for i, pitch_idx in enumerate(strum_order):
        pitch = voicing_pitches[pitch_idx]

        # Calculate start time with strum
        # Distribute notes evenly across strum duration
        if len(strum_order) > 1 and strum_beats > 0:
            strum_position = i / (len(strum_order) - 1) if len(strum_order) > 1 else 0
            note_start = base_start_tick + int(strum_position * strum_ticks)
        else:
            note_start = base_start_tick

        # Apply humanization (clamp to not exceed chord bounds)
        if humanize_beats > 0:
            humanize_offset = rng_humanize.randint(-humanize_ticks, humanize_ticks)
            note_start += humanize_offset
            # Clamp to chord bounds
            note_start = max(base_start_tick, min(base_start_tick + duration_ticks, note_start))

        # Calculate velocity with intensity and jitter
        base_velocity = int(100 * intensity)
        if chord_event.velocity_jitter > 0:
            velocity_jitter = rng_velocity.randint(
                -chord_event.velocity_jitter, chord_event.velocity_jitter
            )
            base_velocity += velocity_jitter
        velocity = max(1, min(127, base_velocity))

        notes.append({
            "pitch": pitch,
            "start_tick": note_start,
            "duration_tick": duration_ticks,
            "velocity": velocity,
        })

    return notes


def apply_voicing(
    chord_tones: list[int],
    voicing: str,
    inversion: int,
    voicing_low: int = 48,
    voicing_high: int = 72,
) -> list[int]:
    """Apply voicing preset and inversion to chord tones.

    Args:
        chord_tones: List of MIDI pitches (can be in any octave)
        voicing: Voicing preset ("root", "open", "drop2", "smooth")
        inversion: Inversion number (0=root, 1=first, 2=second)
        voicing_low: Lowest allowed MIDI pitch
        voicing_high: Highest allowed MIDI pitch

    Returns:
        List of MIDI pitches within range
    """
    if not chord_tones:
        return []

    # Normalize to pitch classes
    pitch_classes = sorted({p % 12 for p in chord_tones})

    # Apply inversion (rotate pitch classes)
    for _ in range(inversion):
        if pitch_classes:
            pitch_classes = pitch_classes[1:] + [pitch_classes[0] + 12]

    # Find octave range
    min_octave = (voicing_low - max(pitch_classes)) // 12
    max_octave = (voicing_high - min(pitch_classes)) // 12

    if min_octave > max_octave:
        octave = (voicing_low + voicing_high) // 24
        return [p + octave * 12 for p in pitch_classes if voicing_low <= p + octave * 12 <= voicing_high]

    # Apply voicing preset
    if voicing == "open":
        # Spread across wider range
        if len(pitch_classes) >= 3:
            # Root, 5th, 3rd spread
            octave_root = min_octave
            octave_third = min_octave + 1
            octave_fifth = min_octave + 1
            result = []
            if len(pitch_classes) > 0:
                result.append(pitch_classes[0] + octave_root * 12)
            if len(pitch_classes) > 1:
                result.append(pitch_classes[1] + octave_third * 12)
            if len(pitch_classes) > 2:
                result.append(pitch_classes[2] + octave_fifth * 12)
            return [p for p in result if voicing_low <= p <= voicing_high]

    elif voicing == "drop2":
        # Drop 2nd voice down an octave
        if len(pitch_classes) >= 2:
            result = []
            for i, pc in enumerate(pitch_classes):
                if i == 1:  # 2nd voice
                    octave = min_octave
                else:
                    octave = min_octave + 1
                pitch = pc + octave * 12
                if voicing_low <= pitch <= voicing_high:
                    result.append(pitch)
            return result

    # Default: root position in middle octave
    octave = (min_octave + max_octave) // 2
    return [p + octave * 12 for p in pitch_classes if voicing_low <= p + octave * 12 <= voicing_high]

